### PR TITLE
IRC, iterated register coalescing (draft implementation)

### PR DESCRIPTION
**Pull request by xclerc**
# Overview

The code is mainly a naive/direct implementation/translation of the
paper (https://dl.acm.org/doi/abs/10.1145/229542.229546), except
the `freeze_moves` function which is actually taken from the book
(https://www.cs.princeton.edu/~appel/modern/ml/).
Small/trivial additions to the paper include:
- handling of register classes, by using per-class values for `k`
  (an early version was adding interference between registers of
   different classes, which obviously results in a denser graph)
- use of `Proc.destroyed_at_xyz`
- handling of written-but-never-read temporaries
- a "naive" splitting, whose objective is to introduce a number of
  spill/reload moves linear in the number of destruction points, rather
  than linear in the number of read/writes when a temporary is spilled
- if there are destruction points in the function, `rewrite` is called
  before `main` (indeed, otherwise the first round of the loop will
  basically *discover* that the values live at the destruction points
  needs to be spilled, then call `rewrite` before launching a second
  round -- we hence essentially save one round by calling `rewrite`
  beforehand when we know temporaries will be spilled)
- *lots of log statements*

# Code

New modules:
- `Cfg_regalloc_utils` contains generic utility functions;
- `Cfg_irc_utils` contains more utility functions, used specifically by
  the IRC allocator;
- `Cfg_irc_state` defines the state of the IRC allocator; it is abstract
  so that it is (relatively) easy to change the representation (*e.g.*
  of the working sets/lists) with no impact on the core of the allocator;
- `Cfg_irc_split` implements a pre-processing phase that performs
  a kind of "naive" or "pseudo" splitting at destruction points;
  essentially, if a register is live at a destruction point (currently only
  calls), we rewrite:
  ```
  destruction_point
  ```
  to:
  ```
  t' <- t
  destruction_point[t/t']
  t <- t'
  ```
- `Cfg_irc` contains the core of the algorithm.

Modified modules:
- `Asmgen` is tweaked to allow the selection of the register allocator,
  and collect metrics;
- `Cfg` is augmented with utility functions;
- `Cfg_intf` is tweaked so that two fields of `instruction`, namely
  `arg` and `res`, are now mutable, and a new field is added to store
  the work list the instruction belongs to;
- `Cfg_dataflow` and `Cfg_liveness` are tweaked so that the transfer
  function take an additional parameter indicating whether the instruction
  can raise;
- `Cfg_equivalence` is tweaked to disable the comparison of live set on
  certain instructions;
- `Emit` is augmented to collect metrics (number of instructions, moves,
  spills and reloads);
- `Reg` is tweaked so that a register now holds its class, its work list,
  its color, and its alias.

# Use

At this point, the register allocators are enabled and controlled by
environment variables. `REGISTER_ALLOCATOR` selects the register
allocator:
- if the variable is unset, the existing register allocator will be used;
- if the variable is set to `irc`, then the IRC allocator will be used
  (note that case is ignored).
  
All allocators can use the `REGALLOC_STATS` variable to enable the production
of metrics. When the variable is set, its value is interpreted as the path to
a directory where quasi-CSV files will be produced. One file will be produced
by compiler invocation, and its base name will be the MD5 digest of the
command-line parameter. The first line of the file will contain the command-line
arguments, while the other lines are CSV contents. Each CSV line contains the
various metrics for a given function.
  
The IRC allocator can be further controlled through the following variables:

- `IRC_VERBOSE` to output *a lot* of debug statements;
- `IRC_INVARIANTS` to check invariants at each step of the process;
- `IRC_SPLIT` to select the kind of splitting to apply before register
  allocation proper (`off` for no splitting and `naive` for the pseudo-
  or naive splitting described above);
- `IRC_SPILLING_HEURISTICS` to select the heuristics to use when
  choosing a register to spill (`set-choose` to randomly choose a
  register and `flat-uses` to choose the least-used register).

# Generated code

The current version of IRC is worse than the current register allocator
in at least two areas: spilling code and memory operands.

The original article will produce a spill (resp. reload) for each write (resp. read)
use of a register that is being spilled (this is the behaviour with `IRC_SPLIT` set
to `off`). The naive splitting will produce a spill/reload for each destruction
point of a register that is being spilled (this is the behaviour with `IRC_SPLIT`
set to `naive`). Both are worse than the split pass implemented by the existing
register allocator.
  
The logic from the `Reload` module is not reused / reimplemented, which means
that all operands are in hardware register and that we do not take advantage of
the fact that some operations can accept one operand on the stack. Put
differently, the current implementation always reloads the spilled value into a
register before the operation while the existing register allocator can use the
value directly from the stack.

# Tests and benchmarks

The IRC allocator has been successfully tested (building the distribution,
and running the test suite from upsteam) with closure, flambda and flambda2.

More numbers will be collected in the upcoming weeks, but here are some
rough numbers collected while building the standard library:

| splitting | one round | all |
| --- | --- | --- |
| off | 0.81-0.86 | 1.10-1.11 |
| naive | 1.18-1.20 | 3.65-3.70 |

(the values above are the _irc_time / upstream_time_ ratios)
("one round" means that the value is about the functions for which
the register allocator needs only one round, while "all" is about all
functions)

Still on the standard library, if we break down per function and look
at the statistics, we get:

| splitting | stat| one round | all |
| --- | --- | --- | --- |
| off | min | 0.28-0.32 | 0.20-0.38 |
| off | mean | 0.86-0.89 | 0.97-0.99 |
| off | max | 11.60-15.49 | 11.60-15.49 |
| naive | min | 0.20-0.26 | 0.18-0.22 |
| naive | mean | 0.99-1.00 | 1.09-1.10 |
| naive | max | 19.11-29.7 | 31.53-31.99 |

The total build times for the compiler distribution (thus not only
measuring register allocation) are:

- with existing allocator: 1
- with IRC and no splitting: 1.04
- with IRC and naive split: 1.05

# TO DO

- check whether the `Reg.clas` field is really useful
- check whether storing the liveness in the instructions would
  make a material difference
- work out what could be gained from other representation (both
  for the IRC state and CFG values)
- implement the "hierarchical" heuristics for spilling
- implement proper spilling
- implement the equivalent of (or simply reuse) `Reload` so that
  we do not always reload a spilled value into a register before an
  operation that can accept an operand from the stack

### CODE REVIEW

**Review by gretay-js — APPROVED** [review]
I reviewed the code of this PR, alongside the Tiger book, it looks good to me. I think this PR should be merged as is, after it is rebased and CI tests for cfg reenabled (they should work now, since PR#656 turned off equivalence checking by default). Further comments should be addressed in separate PRs.

I have a few  minor questions in the diff of the PR, but they shouldn't block this from being merged.  Additionaly, suggestions / questions for later:

- Create a separate directory for regalloc (either under cfg or next to it, not sure)
- For testing, can we run a version of reload pass once to guarantee that the constrains that Emit relies on are satisfied after IRC. This requires reimplementing the logic of reload on cfg.
- Separate the stats from Emit. Add a separate pass.
- Stats: spills in upstream are either Ispill or Imove with res on
  stack. Similarly for reloads. It may be useful to separate Imove to
  spill and reload.
- What will ocaml-ir do if a pass is missing. e.g, Compiler_hooks.Mach_spill is never called in the IRC pipeline?
- Is it possible to refactor Asmgen.compile_fundecl to move stats out of it into a helper function, or somehow tidy up the code, it's a bit hard to follow (maybe it'll go away).
- In cfg_liveness: has_an_exceptional_successor why is it needed?
- document magic number "16" in cfg_dataflow
- cfg_equivalence,  check_live: can you please add a comment why live is
   not checked for poptrap/pushtrap/prologue. These instructions don't exists in Mach. Do they get wrong made-up values somehow in Linear_to_cfg or Cfgize?
- if we have mutable fields in Cfg.instruction we should make live mutable too again and see if it helps. 
- remove_prologue_if_not_required : why does it need to be added on always, instead of adding it on after regalloc and only if required? (there was something about it being after debug related instructions).
- cfg_regalloc_utils.ml: maybe rename Fetch -> Load
- Stats separate module
- Cfg_regalloc_utils Instruction: something is wrong in our Cfg interface if we need this.
- Cfg_state: common interface to the various worklists?
- Is there a check that there is no edge between two Reg.t that are precolored in different colors?
- Assert the item is in the expected worklist before removing it (e.g., remove_spill_work_list).

**Comment by poechsel** [comment]
> What will ocaml-ir do if a pass is missing. e.g, Compiler_hooks.Mach_spill is never called in the IRC pipeline?

Nothing apart from outputting an error message. I don't think many people rely on OCaml-IR supporting Mach_spill so I'm fine with IRC skipping Mach_spill.

**Review by mshinwell — APPROVED** [review]
Approving the changes to `dune` only.

**Review by gretay-js — APPROVED** [review]

File: backend/reg.ml [hunk 1]
```diff
@@ -60,8 +88,14 @@ and stack_location =
 
 type reg = t
 
+(* CR xclerc for xclerc: amd64 *)
+let clas_of_typ = function
```
[the remark this answers is not in the document]
Comment (xclerc) [reply backend/reg.ml]: Would create a circular dependency -- remove the `clas` field.

File: backend/cfg/cfg_liveness.ml [hunk 2]
```diff
@@ -56,8 +61,13 @@ struct
...
       else
         let across = Reg.diff_set_array before instr.res in
+        let can_raise =
+          (* CR xclerc for xclerc: used to be `Cfg.can_raise_basic
+             instr.desc`. *)
+          has_an_exceptional_successor
+        in
         let across =
-          if Cfg.can_raise_basic instr.desc && instr.stack_offset > 0
+          if can_raise && instr.stack_offset > 0
```
[the remark this answers is not in the document]
Comment (xclerc) [reply backend/cfg/cfg_liveness.ml]: Disappeared in the rebase.

File: backend/cfg/merge_straightline_blocks.ml [hunk 3]
```diff
@@ -65,7 +65,8 @@ let rec merge_blocks (modified : bool) (cfg_with_layout : Cfg_with_layout.t) :
              && (not (Label.equal b1_label b2_label))
              && List.compare_length_with b2_predecessors 1 = 0
              && Cfg.is_pure_terminator b1_block.terminator.desc
-             && not b1_block.can_raise
+             && Label.Set.is_empty
```
[the remark this answers is not in the document]
Comment (xclerc) [reply backend/cfg/merge_straightline_blocks.ml]: Disappeared in the rebase.

File: backend/cfg/cfg_regalloc_utils.mli [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+  module IdSet : MoreLabels.Set.S with type elt = id
+
+  module IdMap : MoreLabels.Map.S with type key = id
+end
+
+val destroyed_at_basic : Cfg.basic -> Reg.t array
+
+val destroyed_at_terminator : Cfg.terminator -> Reg.t array
+
+val iter_instructions :
```
[the remark this answers is not in the document]
Comment (xclerc) [reply backend/cfg/cfg_regalloc_utils.mli]: Moved.

File: backend/cfg/cfg_regalloc_utils.mli [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+val collect_cfg_infos : Cfg_with_layout.t -> cfg_infos
+
+type liveness = Cfg_liveness.Liveness.domain Cfg_dataflow.Instr.Tbl.t
+
+val liveness_analysis : Cfg_with_layout.t -> liveness
+
+module Move : sig
+  type t =
+    | Plain
+    | Fetch
```
[the remark this answers is not in the document]
Comment (xclerc) [reply backend/cfg/cfg_regalloc_utils.mli]: Renamed.

File: backend/cfg/cfg_irc_state.ml [hunk 6]
```diff
@@ -0,0 +1,617 @@
...
+  let n = reg in
+  iter_node_moves state n ~f:(fun m ->
+      match m.irc_work_list with
+      | Active ->
+        m.irc_work_list <- Work_list;
+        state.active_moves <- remove_instr m state.active_moves;
+        state.work_list_moves <- m :: state.work_list_moves
+      | _ -> ())
+
+let[@inline] enable_moves_set state regset =
```
[the remark this answers is not in the document]
Comment (xclerc) [reply backend/cfg/cfg_irc_state.ml]: Removed.

File: backend/cfg/cfg_regalloc_utils.mli [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    id:Instruction.id ->
+    copy:_ Cfg.instruction ->
+    from:Reg.t ->
+    to_:Reg.t ->
+    Instruction.t
+
+  val to_string : t -> string
+end
+
+val same_reg : Reg.t -> Reg.t -> bool
```
[the remark this answers is not in the document]
Comment (xclerc) [reply backend/cfg/cfg_regalloc_utils.mli:56]: Moved `same_reg`; `same_reg_class` cannot, it would
create a circular dependency.

File: backend/cfg/cfg_irc.ml [hunk 8]
```diff
@@ -0,0 +1,661 @@
...
+    | Naive -> (
+      match naive_split_points cfg_with_layout with
+      | [] -> []
+      | _ :: _ as split_points ->
+        let liveness = liveness_analysis cfg_with_layout in
+        naive_split_cfg state cfg_with_layout split_points liveness)
+  in
+  let spilling_because_split_or_unused : Reg.t list =
+    Reg.Set.fold
+      (fun reg acc -> if Reg.Set.mem reg arg then acc else reg :: acc)
```
[the remark this answers is not in the document]
Comment (gretay-js) [reply backend/cfg/cfg_irc.ml:573]: Computing diff explicitly is more costly. Other questions will be addressed separately.

File: backend/cfg/cfg_irc_state.ml [hunk 9]
```diff
@@ -0,0 +1,617 @@
...
+  let d = reg.Reg.degree in
+  if d = Degree.infinite
+  then ()
+  else begin
+    reg.Reg.degree <- pred d;
+    if Int.equal d (k reg)
+    then begin
+      enable_moves_one state reg;
+      iter_adjacent state reg ~f:(fun r -> enable_moves_one state r);
+      reg.Reg.irc_work_list <- Reg.Unknown_list;
```
[the remark this answers is not in the document]
Comment (gretay-js) [reply backend/cfg/cfg_irc_state.ml:400]: will be addressed in the worklist optimization patch

**Review by gretay-js — APPROVED** [review]
